File: capture-pwa/src/gates/captureArmEvaluator.ts
import {
  createInitialCaptureArmState,
  type CaptureArmEvaluation,
  type CaptureArmInputs,
  type CaptureArmState,
} from './captureArmTypes';

/**
 * Pure function, same contract as evaluateSmartFrame: deterministic
 * given (input, prevState), no DOM/timer access. `holdReached` only
 * matters while idle (it's a one-shot edge from evaluateSmartFrame's
 * captureTriggered, which by construction can't re-fire mid-count
 * without holdCount resetting first); `allPassed` is read continuously
 * while counting, and any false cancels immediately.
 */
export function evaluateCaptureArm(
  input: CaptureArmInputs,
  prevState: CaptureArmState,
): CaptureArmEvaluation {
  const { holdReached, allPassed, nowMs, durationMs } = input;

  if (prevState.phase === 'idle') {
    if (!holdReached) {
      return {
        phase: 'idle',
        displayTick: null,
        tickJustChanged: false,
        fireNow: false,
        justCanceled: false,
        state: prevState,
      };
    }
    return evaluateCounting({ phase: 'counting', armedSince: nowMs, lastAnnouncedTick: -1 }, nowMs, durationMs);
  }

  if (!allPassed) {
    return {
      phase: 'idle',
      displayTick: null,
      tickJustChanged: false,
      fireNow: false,
      justCanceled: true,
      state: createInitialCaptureArmState(),
    };
  }
  return evaluateCounting(prevState, nowMs, durationMs);
}

function evaluateCounting(state: CaptureArmState, nowMs: number, durationMs: number): CaptureArmEvaluation {
  const elapsed = nowMs - state.armedSince;
  if (elapsed >= durationMs) {
    return {
      phase: 'idle',
      displayTick: null,
      tickJustChanged: false,
      fireNow: true,
      justCanceled: false,
      state: createInitialCaptureArmState(),
    };
  }

  const remainingMs = Math.max(0, durationMs - elapsed);
  const displayTick = Math.max(1, Math.ceil(remainingMs / 1000));
  const tickJustChanged = displayTick !== state.lastAnnouncedTick;
  const nextState: CaptureArmState = {
    phase: 'counting',
    armedSince: state.armedSince,
    lastAnnouncedTick: displayTick,
  };

  return {
    phase: 'counting',
    displayTick,
    tickJustChanged,
    fireNow: false,
    justCanceled: false,
    state: nextState,
  };
}

File: capture-android/app/src/main/java/com/gavan/capture/gates/CaptureArmEvaluator.kt
package com.gavan.capture.gates

import kotlin.math.ceil
import kotlin.math.max

/**
 * Pure function, same contract as evaluateSmartFrame: deterministic
 * given (input, prevState), no timer access of its own. Mirrors
 * capture-pwa/src/gates/captureArmEvaluator.ts exactly.
 */
fun evaluateCaptureArm(input: CaptureArmInputs, prevState: CaptureArmState): CaptureArmEvaluation {
    val (holdReached, allPassed, nowMs, durationMs) = input

    if (prevState.phase == CaptureArmPhase.IDLE) {
        if (!holdReached) {
            return CaptureArmEvaluation(
                phase = CaptureArmPhase.IDLE,
                displayTick = null,
                tickJustChanged = false,
                fireNow = false,
                justCanceled = false,
                state = prevState,
            )
        }
        return evaluateCounting(
            CaptureArmState(phase = CaptureArmPhase.COUNTING, armedSince = nowMs, lastAnnouncedTick = -1),
            nowMs,
            durationMs,
        )
    }

    if (!allPassed) {
        return CaptureArmEvaluation(
            phase = CaptureArmPhase.IDLE,
            displayTick = null,
            tickJustChanged = false,
            fireNow = false,
            justCanceled = true,
            state = createInitialCaptureArmState(),
        )
    }
    return evaluateCounting(prevState, nowMs, durationMs)
}

private fun evaluateCounting(state: CaptureArmState, nowMs: Double, durationMs: Long): CaptureArmEvaluation {
    val elapsed = nowMs - state.armedSince
    if (elapsed >= durationMs) {
        return CaptureArmEvaluation(
            phase = CaptureArmPhase.IDLE,
            displayTick = null,
            tickJustChanged = false,
            fireNow = true,
            justCanceled = false,
            state = createInitialCaptureArmState(),
        )
    }

    val remainingMs = max(0.0, durationMs - elapsed)
    val displayTick = max(1, ceil(remainingMs / 1000.0).toInt())
    val tickJustChanged = displayTick != state.lastAnnouncedTick
    val nextState = CaptureArmState(
        phase = CaptureArmPhase.COUNTING,
        armedSince = state.armedSince,
        lastAnnouncedTick = displayTick,
    )

    return CaptureArmEvaluation(
        phase = CaptureArmPhase.COUNTING,
        displayTick = displayTick,
        tickJustChanged = tickJustChanged,
        fireNow = false,
        justCanceled = false,
        state = nextState,
    )
}
